src/mpi/comm/comm_split_type.c

```diff
@@ -73,6 +70,142 @@ int MPIR_Comm_split_type_node(MPIR_Comm * comm_ptr, int key, MPIR_Comm ** newcom
     goto fn_exit;
 }
 
+#ifdef HAVE_HWLOC
+
+struct shmem_info_table {
+    const char *val;
+    hwloc_obj_type_t hwloc_obj_type;
+    int depth;
+};
+
+static struct shmem_info_table shmem_info[] = {
+    {"socket", HWLOC_OBJ_PACKAGE, -1},
+    {"package", HWLOC_OBJ_PACKAGE, -1},
+    {"numanode", HWLOC_OBJ_NUMANODE, -1},
+    {"core", HWLOC_OBJ_CORE, -1},
+    {"hwthread", HWLOC_OBJ_PU, -1},
+    {"l1cache", HWLOC_OBJ_CACHE, 1},
+    {"l2cache", HWLOC_OBJ_CACHE, 2},
+    {"l3cache", HWLOC_OBJ_CACHE, 3},
```

Comment:
Should these be split into data, instruction and unified caches, and default to data/unified caches if unspecified?